src/System.Management.Automation/engine/hostifaces/PSTask.cs

```diff
@@ -0,0 +1,1343 @@
+        }
+
+        /// <summary>
+        /// Initializes instance of PSTaskBase.
+        /// </summary>
+        /// <param name="scriptBlock">Script block to run</param>
+        /// <param name="usingValuesMap">Using variable values passed to script block</param>
+        /// <param name="dollarUnderbar">Dollar underbar variable value</param>
+        protected PSTaskBase(
+            ScriptBlock scriptBlock,
+            Dictionary<string, object> usingValuesMap,
+            object dollarUnderbar) : this()
+        {
+            _scriptBlockToRun = scriptBlock;
+            _usingValuesMap = usingValuesMap;
+            _dollarUnderbar = dollarUnderbar;
+        }
+
+        #endregion
+
+        #region Abstract Methods
+
+        /// <summary>
+        /// Initialize PowerShell object.
+        /// </summary>
+        protected abstract void InitializePowershell();
+
+        #endregion
+
+        #region IDisposable
+
+        /// <summary>
+        /// Dispose PSTaskBase instance.
+        /// </summary>
+        public void Dispose()
+        {
+            _runspace.Dispose();
+            _powershell.Dispose();
+            _output.Dispose();
+        }
+
+        #endregion
+
+        #region Public Methods
+
+        /// <summary>
+        /// Start task.
+        /// </summary>
+        public void Start()
+        {
+            if (_powershell != null)
+            {
+                Dbg.Assert(false, "A PSTask can be started only once.");
+                return;
+            }
+
+            // Create and open Runspace for this task to run in
+            var iss = InitialSessionState.CreateDefault2();
+            iss.LanguageMode = (SystemPolicy.GetSystemLockdownPolicy() == SystemEnforcementMode.Enforce) 
+                ? PSLanguageMode.ConstrainedLanguage : PSLanguageMode.FullLanguage;
+            _runspace = RunspaceFactory.CreateRunspace(iss);
+            _runspace.Name = string.Format(CultureInfo.InvariantCulture, "{0}:{1}", RunspaceName, s_taskId);
+            _runspace.Open();
+
+            // Create the PowerShell command pipeline for the provided script block
+            // The script will run on the provided Runspace in a new thread by default
+            _powershell = PowerShell.Create();
+            _powershell.Runspace = _runspace;
+
+            // Initialize PowerShell object data streams and event handlers
+            _output = new PSDataCollection<PSObject>();
+            InitializePowershell();
+
+            // Start the script running in a new thread
+            _powershell.AddScript(_scriptBlockToRun.ToString());
+            _powershell.Commands.Commands[0].DollarUnderbar = _dollarUnderbar;
+            if (_usingValuesMap != null && _usingValuesMap.Count > 0)
+            {
+                _powershell.AddParameter(Parser.VERBATIM_ARGUMENT, _usingValuesMap);
+            }
+            _powershell.BeginInvoke<object, PSObject>(null, _output);
```

Comment:
please use named parameters

```suggestion
            _powershell.BeginInvoke<object, PSObject>(input:null, output:_output);
```